### PR TITLE
Nested expand field

### PR DESCRIPTION
**Pull request by mdmunir**
#8061 #6844

### CODE REVIEW

**Comment by bazaglia** [comment]
Tried locally and it works. Also, looking good! Thanks @mdmunir. Hope it gets merged soon :)

File: framework/helpers/BaseArrayHelper.php [hunk 1]
```diff
@@ -56,13 +56,20 @@ class BaseArrayHelper
      * @param boolean $recursive whether to recursively converts properties which are objects into arrays.
      * @return array the array representation of the object
      */
-    public static function toArray($object, $properties = [], $recursive = true)
+    public static function toArray($object, $properties = [], $recursive = true, $expands = [])
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php]: Method signature changed. Can't merge it in 2.0.x. 2.1 only.

**Comment by tunecino** [comment]
nice feature @mdmunir but is there a way to control authorization ? I mean when used to navigate where we are not authorized to like: `/item?expand=cart.owner.account`

**Comment by tunecino** [comment]
otherwise I think it should be disabled by default to be implemented after unsetting related fields within the checkAccess method.

**Comment by fernandezekiel** [comment]
@tunecino the problem with this kind of design is that the fields are declared inside a model instance...
i would like to see somebody implement another layer for this one to implement in controller level wherein it can filter by user's credentials which expand it can see or not...

it can still be done via the model to some degree...
i just have a subclass of my main model that does the check

**Comment by tunecino** [comment]
I think you are right @fernandezekiel.

Other thing that may help solving this could be by first solving Nested Urls. then we can also have something that works like `scenarios` in `model::rules` to control fields expanding. like when doing for example `GET /carts/2/orders/4` we can easily have pre-setted conditions on fields to auto expand some related relations like the data in the juction_table between both models.

I think I have few ideas on implementing nested urls via a custom UrlRule class. I'll start by that. maybe an extension. if it works we can implement similar thing to the framework. if it dosent, who knows, maybe we help influencing others on figuring out a good solution.

**Comment by fernandezekiel** [comment]
the issue is related here though
https://github.com/yiisoft/yii2/issues/7098

**Comment by fernandezekiel** [comment]
i have some implementation of this one already, although it kind of makes me re-declare the mapped return fields in the controller... 
i can make the usage also that if it's not declared in the expandDependency that it will not be in the response...
i also use this for eager loading or callbacks for the query object if a field was requested in the expand
```php

    /**
     * usages:
     * 1. declare the expand field itself as the array value without a string pair, example: location
     * 2. declare the expand field by key value pair , the key will be the expand field requested and the
     * value can be a string or an array of the needed relations to be eager loaded, example: 'created_user' => 'createdUser',
     * 'is_deletable' => ['productAchPv', 'productGateway']
     * 3. declare the expand field wherein the other value is a callable that accepts query as a parameter and will return
     * the query object
     * ~~~
     * return [
     *      'location', 'created_user' => 'createdUser',
     *      'is_deletable' => ['recurrings', 'projects'],
     *      'special_field' => function ($query) {
     *          return $query;
     *      }
     * ];
     * ~~~
     * @return array
     */
    public function expandDependency()
    {
        return [];
    }
```

**Comment by samdark** [comment]
Won't make it into 2.0.

**Pull request closed without merging**